### PR TITLE
Improved performance of the crawler

### LINKED ISSUE #9041: Improve the performance of the crawler
I tried to understand what's causing the crawler to be so slow. It's odd because running it for manual tests (215 in CKE5 itself) should take maybe 215\*15s. 15 is enough to serve all files and execute the editor, right? Unfortunately, I'm at 52% done and testing the next test takes over 1 minute. This is odd.

What's also odd is this:

Why is chrome taking 103% of my CPU while the Idle is 86.7%. It somehow does not add up, but perhaps it's only how macOS shows the load.

Anyway, 1 minute per test is odd. And it seems like it slows down with time - every next test takes more time.

I found the crawler super useful but the performance aspect makes it hard to use at times. I'd love if we find the solution as it seems that it's not perhaps related to using one thread (although, spawning more may help significantly). It seems like something else is not working well.

E.g. how do you open and close the new test?

_Originally posted by @Reinmar in https://github.com/ckeditor/ckeditor5/issues/8800#issuecomment-779925825_

### PR DESCRIPTION
**Pull request by psmyrek**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/contributing/git-commit-message-convention.html))

Other: Improved performance of the crawler. Closes #9041.

---

### Additional information

Notable changes:

* New CLI argument `-c` (`--concurrency`) to define number of concurrent pages (browser tabs) to be used during crawling.

* New types of errors, that appear while navigating to pages, are detected now:
  
  * there's an SSL error (e.g. in case of self-signed certificates),
  * target URL is invalid,
  * the timeout is exceeded during navigation (e.g. in case of an infinite loop in the code),
  * the remote server does not respond or is unreachable,
  * the main resource failed to load.

* Each link is opened in a dedicated Puppeteer's page instance, which is a kind of sandbox.

* Performance has been significantly improved, compared to the previous version of the crawler:

  | | Scanning the docs | Scanning manual tests |
  | --- | :---: | :---: |
  | Before | 7 min 30 s | 8 min 10 s |
  | After | **1 min 30 s** | **1 min 40 s** |

  After many experiments, it seems that the optimal number of simultaneously opened pages in the headless browser is 10, so 10 is the default value used in docs and manual tests scanning (in `yarn docs:verify` and `yarn manual:verify` scripts, respectively).

### CODE REVIEW

**Review by pomek — COMMENTED** [review]
I executed the new version of crawler with the documentation. I ended with a few errors:

```
  ❌ HTTP code 404 for http://fake.ckeditor.com:8080/index.html

    …occurred 1 time(s) on the following page(s):
    ➥  about:blank
```

(1) The docs builder will not build the page so we need to ignore this error.

```
  ❌ HTTP code 404 for http://fake.ckeditor.com:8080/docs/ckeditor5/19.0.0/framework/guides/contributing/development-environment.html

    …occurred 1 time(s) on the following page(s):
    ➥  about:blank
```

Source of the error: https://github.com/ckeditor/ckeditor5/blob/63ffdb0c8eefc9f7cc649814aa397c4f244ed595/docs/framework/guides/contributing/development-environment.md#L20-L22.
(2) As above, we need to ignore this error as well. The documentation builder builds only the `latest` version of our docs and the past builds are not available here.

```
  ❌ HTTP code 404 for http://fake.ckeditor.com:8080/ckeditor5/25.0.0/features/image.png

    …occurred 1 time(s) on the following page(s):
    ➥  about:blank

  ❌ HTTP code 404 for http://fake.ckeditor.com:8080/ckeditor5/latest/features/image.png

    …occurred 1 time(s) on the following page(s):
    ➥  about:blank
✨  Done in 123.23s.
```

(3) This one was tricky. It comes from the Link guide.

image

I was thinking that we should ignore HTML produced by the editor itself (editing view). But in some cases, it makes sense to check whether the links in the editor's content are valid. So, we should improve the guide itself.

(4) As you can see, the last errors come from the same page that can be visited using two URLs: exact version (`/25.0.0/`) and the latest version (`/latest/`). It means that our documentation will be checked twice (1st time for exact version, 2nd for the latest).

We could avoid this by a simple change in [the crawler configuration](https://github.com/ckeditor/ckeditor5/blob/4b11de7ef1eb5258d820236c0d34b7a3aa773286/scripts/web-crawler.js#L440):

```diff
- const defaultOptionsForDocs = minimist( '-u http://fake.ckeditor.com:8080/ckeditor5/ -e /api/,/assets/ -c 10'.split( ' ' ), config );
+ const defaultOptionsForDocs = minimist( '-u http://fake.ckeditor.com:8080/ckeditor5/latest/ -e /api/,/assets/ -c 10'.split( ' ' ), config );
```

Other than that, I'd change the default number of browsers tab that is using for crawling. As for, it's 10. While crawling checking pages, my machine freezes and I'm not able to do anything else. It should not work this way.

(5) I propose to use half of the available threads on the machine where the crawler is being called.

Using the concurrency option set to `4`, the documentation and manual tests require 3 mins (per item) to be checked. It still a good time.

---

- (1) and (2) could be handled as follow-ups when working on https://github.com/ckeditor/ckeditor5/issues/9062 or https://github.com/ckeditor/ckeditor5/issues/8800. 
- (3) could be reported as a follow-up and handled by the docs squad.
- (4) an improvement that we can implement in the PR
- (5) I'd say, it requires changes in the PR.

File: scripts/web-crawler.js [hunk 1]
```diff
@@ -12,6 +12,8 @@ const chalk = require( 'chalk' );
 const ora = require( 'ora' );
 const minimist = require( 'minimist' );
 
+const DEFAULT_TIMEOUT = ( 15 * 1000 );
```
[the remark this answers is not in the document]
Comment (psmyrek) [reply scripts/web-crawler.js]: Done.

File: scripts/web-crawler.js [hunk 2]
```diff
@@ -215,37 +214,82 @@ function getProgressHandler( spinner ) {
...
+ * Duplicated links, which were already collected and enqueued, are skipped to avoid loops. Explicitly excluded links
+ * are also skipped.
  *
- * @param {Object} page The page instance from Puppeteer.
+ * @param {Object} browser The headless browser instance from Puppeteer.
  * @param {Object} data All data needed for crawling the links.
  * @param {String} data.host The host from the initial page URL.
- * @param {String} data.link The link to crawl.
- * @param {Array.<String>} data.foundLinks An array of links, which have been already discovered.
+ * @param {Array.<Object>} data.linksQueue An array of link to crawl. Each item has `url` and `depth` keys.
```
[the remark this answers is not in the document]
Comment (psmyrek) [reply scripts/web-crawler.js]: Done.

File: scripts/web-crawler.js [hunk 3]
```diff
@@ -364,8 +437,8 @@ function parseArguments( args ) {
 		throw new Error( 'Mutually exclusive --docs and --manual arguments.' );
 	}
 
-	const defaultOptionsForDocs = minimist( '-u http://fake.ckeditor.com:8080/ckeditor5/ -e /api/,/assets/'.split( ' ' ), config );
-	const defaultOptionsForManual = minimist( '-u http://localhost:8125/ -d 1'.split( ' ' ), config );
+	const defaultOptionsForDocs = minimist( '-u http://fake.ckeditor.com:8080/ckeditor5/ -e /api/,/assets/ -c 10'.split( ' ' ), config );
+	const defaultOptionsForManual = minimist( '-u http://localhost:8125/ -d 1 -c 10'.split( ' ' ), config );
```
[the remark this answers is not in the document]
Comment (psmyrek) [reply scripts/web-crawler.js]: Done.

**Comment by psmyrek** [comment]
Thank you very much for such a detailed review, it was very helpful. As you suggested, some of your notes will be handled in other tasks. The issue with missing `image.png` is already reported: https://github.com/ckeditor/ckeditor5/issues/8778

**Review by pomek — APPROVED** [review]
Wow, 77 secs for checking docs.

**Pull request merged**